### ISSUE TITLE
[EX-39] GTD 'review' functionality

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#35]
_Paul Beckingham on 2012-07-22T03:53:31Z says:_ 

 Now that UDAs are here, it is possible to define a new date, called 'review'.  Then an external script could run through all the pending tasks, updating 'review' as it goes.  Perhaps it would call (at)task < IDs > edit@.

This issue is a placeholder for a discussion about how this should be done.
### We are closer than we think..

In recent discussions.. mostly between me and me.. and Paul, the whole post-2.x landscape is coming into focus. 
A lot of work went into taskwarrior to give it capabilities to do cool stuff, 
now it's time to do cool stuff.

I'm going to use the "review process" to illustrate how a number of features-in-development might/can/will be used to review. In the process I hope to illustrate how a few of these future-features will work in concert. There are (positive) usage implications from all angles, good stuff coming down the pipe.

To start with, let's set aside the old term "review script" in favour of the term "review process", this because the action of this-thing-after-that-thing will be easily accomplished with (upcoming?) new internal functions and behaviours, specifically (# ) command separators "++". The "internal pipe" that will allow the construction of complex sequences, including random shell execution calls. The ability to "chain" arbitrary, complex task queries interspersed with arbitrary shell commands (mostly *echo* and *cat*) obviates the need for an external *review* "script" in python or perl or anything else. Layer-after-layer of abstraction, a "review", a complex report, and most other things boil down to a "simple" alias call, like *review-week-by-project*.

The heavy-lifting of the interactivity will be carried by a combination of *task ask* (#1079) *%variables* (#1073) and *VIRTUALS*, (http://taskwarrior.org/boards/6/topics/2414) which will jump onto centre stage as everybodys favourite way to do everything. 

The following stack is worthy of a flow-chart, and I will make one when I can, but with it I will try to indicate the levels of abstraction and the degrees of subtlety available to us, in creating a review. Perhaps amusingly, this description will _not_ include the full GTD-specific ramifications, that's a subsequent layer ;-)

so a typical task review process (and most other complex functions) might consist of

<pre>
* full set of core attributes
* unlimited User Defined Attributes
* rich set of attribute modifiers (attmods)
* full use of regex logic If AND OR GALORE (available at every (?) following step)

   wrapped in

      Complex strings of VIRTUALS; vfilters, vconfigs and vmods, 
      condensing arbitrary-length conditional logic statements and complex modification-sets
      into a set of deceptively simple, modal terms like +DUE, +HOME and +READY.

         wrapped in

            arbitrary-length compound-command strings delimited by "++", allowing any number of 
            task commands and/or shell-calls (echo and cat, etc) to be issued sequentially, 
            accumulating the output into a final result.

               wrapped in

                   command aliases that take the hairiest of that, and calls it by 
                   something logical and easy to remember, like *review-goals*
</pre>
and you have a stack of layers of abstraction limited only by your imagination. 

The stack above is by no means the only way these features could be combined, and in every-day use, different features will be combined at different layers, but this is sort of the conceptual stack I would suggest for us to design complex features around. 

... in progress <save>

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2012-07-22T03:53:31Z
Modified: 2017-01-17T02:59:59Z
```

**Comment by taskwarrior** [comment]
_David Patrick on 2012-07-24T01:19:04Z says:_ 

 How the review process works, as a whole, should be a rollicking discussion. Personally I think reviews should be a predefined set of scripts, that follow a somewhat GTD (tm) path. 

The need for a "review" date, then, is negated by the idea that the review process is pro-active, and will catch what's in the set of queries.

BUT

The need for a "reviewed" date is then essential to govern the frequency of review. 

I know, I know, it's the same thing!.. but it ISN'T ;-) !
I think we'll need special tags (+review), rc settings (review.proj=foo), and a few canned queries, but let's discuss!

how would YOU like to Review?

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2012-07-24T04:09:09Z says:_ 

 With 300 pending tasks to review, there is no way i want to sit down and review them all.  By recording a 'last reviewed' date, an external script could present the tasks, one by one, in order of greatest need, which probably means least-recently-reviewed.  I would then be able to view the task, and either skip it or edit it.  Once edited, I move on to the next task.

It would need to be fast and lightweight, otherwise I'll find it clunky and I won't do it, and therefore miss out on the benefits of review.  I would need to be able to review, say, 20 tasks and the stop because I want to do something else.  Those 20 tasks would be marked as reviewed, and next time I run the review script, those 20 would not show up, because of that 'least-recently-reviewed' criterion above.

I would want to configure the review so that I am never reviewing tasks more often than, say, two weeks.  If I review all 300 tasks, then four days later run the review script again, it should say "You're all set.  Don't waste your time."

As described above, the review script could be a simple Bash scipt that calls "task info", "task edit" and "task modify", having first ensured that the UDA for last_reviewed is set up.

**Comment by taskwarrior** [comment]
_David Patrick on 2012-07-24T18:22:08Z says:_ 

 Paul Beckingham wrote:
> With 300 pending tasks to review, there is no way i want to sit down and review them all.

ok, so let's try describing a sample review sequence, in the vaguest of psudo-code terms, and for the fun of it, let's imagine the hypothetical user to be a heavy taskwarrior user/ software developer.. here's a couple of things I imagine such a user might want to review;

overdue
everything with +review tag that hasn't been seen in [n period]
blockers to active projects
tasks missing parameters
projects with no action
tasks with positive revenue potential
tasks that are algorithmic candidates for deletion
tasks that are actionable in [pet] project



> It would need to be fast and lightweight, otherwise I'll find it clunky and I won't do it, and therefore miss out on the benefits of review.  I would need to be able to review, say, 20 tasks and the stop because I want to do something else.  Those 20 tasks would be marked as reviewed, and next time I run the review script, those 20 would not show up, because of that 'least-recently-reviewed' criterion above.

I think that the review function is going to look a whole lot like the alias functionality. If "review" was an alias for "alias" (ohhh the recursionn) a set of semi-complex commands could be constructed, even calling-out to external scripts, if needed. The only real differences between "review" and "alias" would be a) review command strings are only executed by the (new) review command, and b) that a "review alias" might be enhanced with prompt text, to guide the user towards the intended function of the command, and c) the review function includes a quick-option prompt. ((e)dit, (d)one (D)elete ? etc)

 review.weekly=due:thisweek list
 review.weekly.prompt=These tasks are due this week, do, delay or delegate
 review.weekly.frequency=1w

> 
> I would want to configure the review so that I am never reviewing tasks more often than, say, two weeks.  If I review all 300 tasks, then four days later run the review script again, it should say "You're all set.  Don't waste your time."

yeah, makes sense to me! At the same time I would NOT consider a task-listing as having reviewed a task, only an "info" display should qualify.

> 
> As described above, the review script could be a simple Bash scipt that calls "task info", "task edit" and "task modify", having first ensured that the UDA for last_reviewed is set up.

and ideally made out of as much existing tw functionality as possible.
$0.02 CDN

**Comment by taskwarrior** [comment]
_David Patrick on 2012-07-24T18:43:11Z says:_ 

 and the command

 %tw review

with no parameters, should invoke the whole set, in the order in which they appear in the rc.

**Comment by taskwarrior** [comment]
_Aikido Guy on 2012-07-24T20:14:09Z says:_ 

 Perhaps I missed reading it... but what about showing groups of tasks at a time? Could be faster to review a group of, say, 10 tasks, than just 1 task at a time?

**Comment by taskwarrior** [comment]
_David Patrick on 2012-07-24T20:20:58Z says:_ 

 Aikido Guy wrote:
> Perhaps I missed reading it... but what about showing groups of tasks at a time?

sure, like an info report on multiple targets, that makes sense, I'm just saying that looking at a "tw list" of tasks isn't really a review, ALTHOUGH, it might be reasonable to set t threshold where if there were more than 20 "hits" then the results could be listed with a quick option to expand those results to the info report. 

It seems to come back to the sophistication of the prompt that might follow a review query, does it offer immediate and appropriate action on the results? Y/n/m ?

**Comment by taskwarrior** [comment]
_David Patrick on 2014-03-24T17:00:30Z says:_ 

 Update: Excellent review functionality has been demonstrated in "trev", a perl-script by Fidel Mato, which can be found here: https://github.com/nocejo/trev. Trev may not be "finished" but it is the best example of a tw Review process.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2017-01-17T02:59:59Z says:_ 

 Implemented in tasksh 1.1.0.
